crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -370,6 +231,192 @@ where
     Ok(CheckpointStepResult::CheckpointSigned)
 }
 
+pub async fn checkpoint_sync_process<A>(
+    active_authority: Arc<ActiveAuthority<A>>,
+    timing: &CheckpointProcessControl,
+    metrics: CheckpointMetrics,
+) where
+    A: AuthorityAPI + Send + Sync + 'static + Clone + Reconfigurable,
+{
+    info!("Start active checkpoint sync process.");
+    tokio::time::sleep(timing.long_pause_between_checkpoints).await;
+    let mut last_cert_time = Instant::now();
+
+    loop {
+        let result = checkpoint_sync_step(active_authority.clone(), timing).await;
+        handle_checkpoint_step(
+            result,
+            active_authority.clone(),
+            timing,
+            metrics.clone(),
+            &mut last_cert_time,
+        )
+        .await;
+    }
+}
+
+async fn checkpoint_sync_step<A>(
+    active_authority: Arc<ActiveAuthority<A>>,
+    timing: &CheckpointProcessControl,
+) -> Result<CheckpointStepResult, CheckpointStepError>
+where
+    A: AuthorityAPI + Send + Sync + 'static + Clone + Reconfigurable,
+{
+    let net = active_authority.net.load().deref().clone();
+
+    // (1) Get the latest checkpoint cert from the network.
+    // TODO: This may not work if we are many epochs behind: we won't be able to download
+    // from the current network. We will need to consolidate sync implementation.
+    let highest_checkpoint = get_latest_checkpoint_from_all(
+        net.clone(),
+        timing.extra_time_after_quorum,
+        timing.timeout_until_quorum,
+    )
+    .await
+    .map_err(|err| CheckpointStepError::SyncCheckpointFromQuorumFailed(Box::new(err)))?;
+
+    // (2) Sync to the latest checkpoint, this might take some time.
+    // Its ok nothing else goes on in terms of the active checkpoint logic
+    // while we do sync. We are in any case not in a position to make valuable
+    // proposals.
+    // Safe to unwrap due to check in the main process function.
+    let state_checkpoints = &active_authority.state.checkpoints;
+
+    if let Some(checkpoint) = highest_checkpoint {
+        debug!(
+            "Highest Checkpoint Certificate from the network: {}",
+            checkpoint
+        );
+        // Check if there are more historic checkpoints to catch up with
+        let next_checkpoint = state_checkpoints.lock().next_checkpoint();
+        if next_checkpoint < checkpoint.summary.sequence_number {
```

Comment:
This will make fullnode always one checkpoint behind, right?

Reply:
@lxfind yeah I initially thought this might be the case, but was unsure since this is the same bound that the fullnode runs for its one-time startup sync... I assumed that the reason it is +1 for a validator is because a validator would be participating in the current checkpoint, whereas the fullnode is just syncing already finished checkpoints. Am I misinterpreting these bounds?

Reply:
If you read the comments for `update_latest_checkpoint`, it's because for validators, the latest checkpoint cert could either be from some validator, or formed on-the-fly.
For fullnode, however, we shouldn't be doing the forming checkpoint cert on the fly. It should strictly try to find the latest cert that some validator in the network has.